### PR TITLE
APP-475: [C++] Different behavior and output of Notification Demo on various SDK clients

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Slash32 @GamovCoder

### CODE REVIEW

**Comment by Slash32** [comment]
@kaa-leeroy test this please

File: notification/source/cpp/src/KaaDemo.cpp [hunk 1]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+        kaaClient_(kaaClient), quitFlag_(quitFlag)
+    {}
+
     virtual void onListUpdated(const Topics& topics)
     {
-        std::cout << ("Topic list was updated") << std::endl;
+        std::cout << "Topic list was updated" << std::endl;
         showTopicList(topics);
         optionalTopicsList_ = extractOptionalTopicIds(topics);
+        for (;;) {
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: `while(true)`

File: notification/source/cpp/src/KaaDemo.cpp [hunk 2]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+            std::cout << "Enter topic id to subscribe to" << std::endl;
+            std::cout << "Enter 'quit' to exit" << std::endl;
+            std::cin.clear();
+
+            int topicId;
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: `int64_t`

File: notification/source/cpp/src/KaaDemo.cpp [hunk 3]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+            std::string input;
+            std::cin >> input;
+
+            if (input == "quit") {
+                quitFlag_ = true;
+                return;
+            }
+
+            try {
+                topicId = std::stoi(input);
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: `stoll`

File: notification/source/cpp/src/KaaDemo.cpp [hunk 4]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+            } catch(std::exception& e) {
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: code style

File: notification/source/cpp/src/KaaDemo.cpp [hunk 5]
```diff
@@ -97,25 +148,18 @@ int main()
 
     kaaClient->start();
 
-    TopicListListener topicListListener;
-    kaaClient->addTopicListListener(topicListListener);
+    std::cout << "Notification demo started" << std::endl;
 
-    NotificationListener notificationListener;
-    kaaClient->addNotificationListener(notificationListener);
+    std::atomic_bool quitFlag{false};
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: Use conditional variable.

File: notification/source/cpp/src/KaaDemo.cpp [hunk 6]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
 
 class NotificationListener : public INotificationListener {
 public:
+    NotificationListener(std::shared_ptr<IKaaClient> kaaClient, std::atomic_bool &quitFlag):
+        kaaClient_(kaaClient), quitFlag_(quitFlag)
+    {}
+
     virtual void onNotification(const int64_t topicId, const KaaNotification &notification)
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: btw, remove `virtual`, add `override`

File: notification/source/cpp/src/KaaDemo.cpp [hunk 7]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+
+            try {
+                kaaClient_->subscribeToTopic(topicId);
+                kaaClient_->syncTopicSubscriptions();
+                return;
+            } catch (KaaException &e) {
+                std::cout << (boost::format("Topic '%1%' is unavailable, can't subscribe") % topicId) << std::endl;
+                continue;
+            }
+            std::cout << (boost::format("Subscribed to topic '%1%'") % topicId) << std::endl;
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: Just wonder, what is the case the message is going to be printed?

File: notification/source/cpp/src/KaaDemo.cpp [hunk 8]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+                topicId = std::stoi(input);
+            } catch(std::exception& e) {
+                std::cout << (boost::format("Incorrect topicId: '%1%'") % input) << std::endl;
+                continue;
+            }
+
+            try {
+                kaaClient_->subscribeToTopic(topicId);
+                kaaClient_->syncTopicSubscriptions();
+                return;
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp:135]: So, I can subscribe only once?

File: notification/source/cpp/src/KaaDemo.cpp [hunk 9]
```diff
@@ -63,23 +63,72 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
         showTopicList(topics);
         optionalTopicsList_ = extractOptionalTopicIds(topics);
+        for (;;) {
+            std::cout << "Enter topic id to subscribe to" << std::endl;
+            std::cout << "Enter 'quit' to exit" << std::endl;
+            std::cin.clear();
+
+            int topicId;
+            std::string input;
+            std::cin >> input;
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp:115]: I'd like to subscribe for a several topics.

**Review by ohamov — CHANGES_REQUESTED** [review]

File: notification/source/cpp/src/KaaDemo.cpp (hunk 1 above)
Comment (RostakaGmfun) [reply notification/source/cpp/src/KaaDemo.cpp]: Don't you think it's a matter of preference? Do we have such requirements in code style guide?

File: notification/source/cpp/src/KaaDemo.cpp (hunk 8 above)
Comment (RostakaGmfun) [reply notification/source/cpp/src/KaaDemo.cpp:135]: It appears to be so according to the bundle scenario flow

File: notification/source/cpp/src/KaaDemo.cpp [hunk 10]
```diff
@@ -63,23 +64,80 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+            if (input == "quit") {
+                std::unique_lock<std::mutex> lock(quitMutex_);
+                quitFlag_ = true;
+                doQuit_.notify_one();
+                return;
+            }
+
+            try {
+                topicId = std::stoll(input);
+            } catch(std::exception& e) {
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: code style

File: notification/source/cpp/src/KaaDemo.cpp [hunk 11]
```diff
@@ -63,23 +64,80 @@ static std::string alertTypeToString(const kaa_notification::AlertType alertType
...
+                std::cout << (boost::format("Incorrect topicId: '%1%'") % input) << std::endl;
+                continue;
+            }
+
+            try {
+                kaaClient_->subscribeToTopic(topicId);
+                kaaClient_->syncTopicSubscriptions();
+                std::cout << (boost::format("Subscribed to topic '%1%'") % topicId) << std::endl;
+                return;
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp:135]: Is it allowed to choose only one of the available topics to subscribe for?

File: notification/source/cpp/src/KaaDemo.cpp (hunk 1 above)
Comment (ohamov) [reply notification/source/cpp/src/KaaDemo.cpp]: It's just a nitpick, but `while (true)` is a common way to express an infinite loop.

File: notification/source/cpp/src/KaaDemo.cpp (hunk 11 above)
Comment (RostakaGmfun) [reply notification/source/cpp/src/KaaDemo.cpp:135]: Yes, according to the bundle scenario flow

Comment (ohamov) [reply notification/source/cpp/src/KaaDemo.cpp:135]: Ok

**Review by ohamov — CHANGES_REQUESTED** [review]
I think there is a deadlock now - a topic list can be received before the listeners are added.

**Review by ohamov — APPROVED** [review]
LGTM

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this fully please

**Comment by Slash32** [comment]
@kaa-leeroy test this fully please

**Pull request merged**